### DISCUSSION TITLE
Rendering many Link components slows down shallow routing

### DISCUSSION THREAD

**Discussion by lmatteis** [vercel/next.js#47914] [Help] [answered]
### Summary

I was able to reproduce an example of a page with many `Link` components. When the router is changed shallowly, all the Link components will re-render and slow down the rendering of another component that depends on the router data (such as query params).

Here's the code example https://codesandbox.io/p/sandbox/modern-meadow-ojmxnj?file=%2Fpages%2Findex.js

* Clicking on any button will shallow link to a new query param on the same page.
* An Overlay component simply listens for changes to these query params and renders it
* Notice how long it takes for the Overlay to render the query param

This seems to be because React has to re-render all of the buttons on the page _first_, because they are Link components (and the router changed). With less buttons on the page, the issue goes away (try changing the value to be lower).

Any thoughts? Is there a way to bypass Link from rendering, especially on shallow routing? Or perhaps there's a way in React 18 to prioritize the render of Overlay components before the thousands of Buttons render.

### Example

https://codesandbox.io/p/sandbox/modern-meadow-ojmxnj?file=%2Fpages%2Findex.js

**Comment by icyJoseph** [comment, marked as answer]
## Edit

One way out is to use imperative navigation:

```jsx
function RandomQueryLink({ children }) {
  const random = Math.floor(Math.random() * 1000);
  const href = `/?random=${random}`;
  return (
    <a
      href={href}
      onClick={(event) => {
        event.preventDefault();
        Router.push(href, undefined, { shallow: true });
      }}
      style={{
        padding: "1rem",
        margin: "0.1rem",
        color: "palevioletred",
        background: "papayawhip",
        display: "inline-block",
      }}
    >
      <span>{children}</span>
    </a>
  );
}
```

For this use case, since shallow is used, the prefetching of JS or _next/data is not needed. We also keep links for SEO.

## Next 12 - Old Link Behavior

Worth mentioning that you are running Next 12 though...

I rewrote it so that it is more painful. Note that this makes it 3 times slower than OPs example.

```jsx
import Link from "next/link";
import { useRouter } from "next/router";
import { useMemo, forwardRef } from "react";

const Anchor = forwardRef((props, ref) => {
  console.count("anchor");
  return (
    <a
      {...props}
      style={{
        padding: "1rem",
        margin: "0.1rem",
        color: "palevioletred",
        background: "papayawhip",
        display: "inline-block",
      }}
      ref={ref}
    />
  );
});

function Button({ children }) {
  // console.count("btn");
  return (
    <Link
      href={{
        pathname: "/",
        query: { random: Math.floor(Math.random() * 1000) },
      }}
      shallow={true}
      passHref
    >
      <Anchor>{children}</Anchor>
    </Link>
  );
}
function Overlay() {
  const router = useRouter();
  return (
    <div style={{ position: "fixed", top: 0 }}>
      Overlay: {router.query.random}
    </div>
  );
}

export default function Home({ data }) {
  const links = useMemo(
    () =>
      Array.from({ length: 20000 }, (_, index) => {
        return <Button key={index}>ciao</Button>;
      }),
    []
  );

  return (
    <div>
      <h1>Hello {data}</h1>
      {links}
      <Overlay />
    </div>
  );
}

export function getServerSideProps() {
  return {
    props: { data: "world" },
  };
}

```

Each click will now count up to 40K in Strict Mode, and 20K without Strict Mode.

## Next 13 - New Link Behavior

In next 13, I don't see such aggressive re-rendering though.

```jsx

function Span(props) {
  console.count("span");
  return (
    <span
      {...props}
      style={{
        padding: "1rem",
        margin: "0.1rem",
        color: "palevioletred",
        background: "papayawhip",
        display: "inline-block",
      }}
    />
  );
}

function Button({ children }) {
  // console.count("btn");
  return (
    <Link
      href={{
        pathname: "/",
        query: { random: Math.floor(Math.random() * 1000) },
      }}
      shallow={true}
    >
      <Span>{children}</Span>
    </Link>
  );
}
```

IIRC the new startTransition must be at play somewhere here. Also the appDir has introduced division between the commands, and the queries, but that's not applicable to your case (pages dir).

**Reply by lmatteis** [reply]
Here's the same file in Next 13 and I can't see much difference in speed: https://codesandbox.io/p/sandbox/modern-meadow-ojmxnj?file=%2Fpages%2Findex.js

**Reply by icyJoseph** [reply]
I see, I think my example that made it worse, did really amplify the issue. I see 5-7 seconds delays with it.

With your exact code, in Next 12/13 I see about the same 1-2 seconds delay.

Used x6 CPU slowdown in all cases. 

I saw in the Link code that startTransition is used if the router is appDir router:

```ts
  if (isAppRouter) {
    // @ts-expect-error startTransition exists.
    React.startTransition(navigate)
  } else {
    navigate()
  }
```

Really spiky issue. Dev Mode makes everything even worse, or rather StrictMode. 

I wonder if this sort of issue requires a hack, like, only render the component if it is within a distance of the viewport, aka virtualization, still render links, just, not next/links... 🤔 when the user scrolls nearby, swap the anchor with a next/link? 

Could you try this out in app dir too? Would be curious to see if the division and heavy usage of React 18 there does salvage the situation.

**Reply by lmatteis** [reply]
Ah that startTransition in the code is very telling. I could try app/ directory but right now my next app isn't using it. So I would like to fix this at the pages/ level first. But will look at Link component code to see what's happening here. Perhaps there's a way to control the transition at the consumer level, using useDeferredValue?

**Reply by icyJoseph** [reply]
The issue, I think, is that the context consumed internally by the Link changes and that's triggering 20_000 render passes. Ideally the readable data (pathname, query) would be part of a context, and the actions (push, replace, etc) would be part of another context.

That way, readers only re-render if the readable stuff changes, and the link could be structured so that it only pushes/replaces - though it would be hard to keep the readable data separated.

From the user-land, I guess, you could use a anchor/button and do programatic with the Router instance, aka Router.push, when onClick triggers, to not read the `useRouter` hook:

```jsx
  <a
          href="/blog"
          onClick={(event) => {
            event.preventDefault();
            Router.push('/blog', null, { shallow: true });
          }}
        >
          Blog
        </a>
```

Something like this seems to work

**Reply by lmatteis** [reply]
Yeah I don't mind about the 20k re-renders. I just want the Overlay render to happen immediately and the rest to do it when idle. With React 18 I thought we could split these kind of computations. Conceptually shouldn't wrapping the Link components in Suspense be a way to tell React "I don't need to render these immediately"? I guess I want some sort of `useDeferredValue` but for the Link component.

PS: I tried the Router.push way and indeed it works, but feels a bit hacky since it's not an official API :(

**Reply by icyJoseph** [reply]
Did you try like this? 

```jsx
import Router, { useRouter } from "next/router";

function Button({ children }) {
  const random = Math.floor(Math.random() * 1000);
  const href = `/?random=${random}`;
  return (
    <a
      href={href}
      onClick={(event) => {
        event.preventDefault();
        Router.push(href, null, { shallow: true });
      }}
    >
      <span>{children}</span>
    </a>
  );
}
```

I had been coding in another language, and was a bit mixed up, now I am 100% back on JavaScript. I think the above should do the trick!

**Reply by icyJoseph** [reply]
Well there's this section of the docs: https://nextjs.org/docs/routing/imperatively so yeah :)

**Reply by lmatteis** [reply]
You still need to call `useRouter` there which causes a re-render.

**Reply by icyJoseph** [reply]
Not sure why I thought to use some kind of SSR swap, you can just lay the href and prevent the default behavior.

You'd missing on pre-fetching, but you have shallow and GSSP, so that's not an issue, right?

**Reply by icyJoseph** [reply]
@lmatteis I use the Router instance, not the `useRouter` hook. Sorry lots of spam.

**Reply by icyJoseph** [reply]
Now I must leave, but I tested and it seems to work well with the imperative navigation and the Router instance combined.

- https://codesandbox.io/p/sandbox/hopeful-star-l86kzs?file=%2Fpages%2Findex.js

One last thing, button as children to anchor tag is not allowed in HTML: https://caninclude.glitch.me/caninclude?child=button&parent=a

Happy to keep helping out!

**Reply by lmatteis** [reply]
Right, makes sense. But I'd still rather not do it imperatively. 

I guess conceptually what I can do is call useRouter somewhere in the parent component instead of all the buttons. 

Indeed this seems to work: https://codesandbox.io/p/sandbox/nervous-mccarthy-tk3qq2?file=%2Fpages%2Findex.js

Thanks I think I'll go down this pathway
